rules/aip0215/foreign_type_reference.go

```diff
@@ -0,0 +1,61 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aip0215
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/googleapis/api-linter/lint"
+	"github.com/googleapis/api-linter/rules/internal/utils"
+	"github.com/jhump/protoreflect/desc"
+	"google.golang.org/protobuf/types/descriptorpb"
+)
+
+var foreignTypeReference = &lint.FieldRule{
+	Name: lint.NewRuleName(215, "foreign-type-reference"),
+	OnlyIf: func(fd *desc.FieldDescriptor) bool {
+		return fd.GetType() == descriptorpb.FieldDescriptorProto_TYPE_MESSAGE
+	},
+	LintField: func(fd *desc.FieldDescriptor) []lint.Problem {
+		curPkg := getPackage(fd)
+		if msg := fd.GetMessageType(); msg != nil {
+			msgPkg := getPackage(msg)
+			if !utils.IsCommonProto(fd.GetMessageType().GetFile()) {
+				if curPkg != "" && msgPkg != "" && !isComponentPackage(msgPkg) && curPkg != msgPkg {
+					return []lint.Problem{{
+						Message:    fmt.Sprintf("foreign type referenced, current field in %q message in %q", curPkg, msgPkg),
+						Descriptor: fd,
+					}}
+				}
+			}
+		}
```

Comment:
I'm kind of a dedent, check assertions, exit early kind of person. How do you feel about this approach? Could be further collapsed into related &&'d conditions I suppose, but I just find a deeply nested conditional hard to grok.

I think the suggestion below is on the extreme end, but perhaps we could check some of the already deeply nested conditions earlier e.g. `if curPkg == "" { return nil }` instead of two levels deeper?

```suggestion
    curPkg := getPackage(fd)
    if curPkg == "" {
      return nil
    }
    msg := fd.GetMessageType()
    if msg == nil {
      return nil
    }
    msgPkg := getPackage(msg)
    if msgPkg == "" {
      return nil
    }
    if curPkg == msgPkg {
      return nil
    }
		if utils.IsCommonProto(fd.GetMessageType().GetFile()) {
	    return nil
		}
		if isComponentPackage(msgPkg) {
			return nil
		}
		return []lint.Problem{{
		  Message:    fmt.Sprintf("foreign type referenced, current field in %q message in %q", curPkg, msgPkg),
		  Descriptor: fd,
	  }}
```

Reply:
Refactored this to align with this style.